### DISCUSSION TITLE
toHaveChildren does not work with a web app on Android

### DISCUSSION THREAD

**Discussion by salmankhilji** [webdriverio/webdriverio#7782] [Q&A]
Given the following HTML:

```html
<!DOCTYPE html>
<html>
  <head>
    <title>Hello World</title>
  </head>
  <body style='background-color: wheat'>
    <p>Hello, World</p>
    <ol id="orderedList">
      <li>Coffee</li>
      <li>Tea</li>
      <li>Milk</li>
      <li>Sugar</li>
    </ol>
  </body>
</html>
```

I want to ensure that the `<ol>` element has exactly 4 child elements using the [toHaveChildren](https://webdriver.io/docs/api/expect-webdriverio/#tohavechildren) API. My test works correctly when the given HTML is loaded into desktop Chrome.  However, the API always returns a **single** child (instead of the expected 4 in my case) when the HTML is loaded into an Android `WebView`. i.e.,

The following works correctly against **desktop Chrome**:

  ```js
  it('should get 4 children under ol using toHaveChildren', async () => {
    await browser.url('./');
    const list = await $('#orderedList');
    await expect(list).toHaveChildren(4);
  });
  ```

However, the following returns only a **single** node on **Android**.

  ```js
  const list = await $('//*[@resource-id="orderedList"]');
  await expect(list).toHaveChildren(4);
  ```

## Desktop Chrome Config

```js
exports.config = {
    specs: [
        './test/specs/**/*.js'
    ],
    exclude: [
        './test/specs/**/*-android.js'
    ],
    maxInstances: 1,
    capabilities: [{
        browserName: 'chrome',
        acceptInsecureCerts: true
    }],
    logLevel: 'info',
    bail: 0,
    baseUrl: 'http://localhost:8000',
    waitforTimeout: 10000,
    connectionRetryTimeout: 120000,
    connectionRetryCount: 3,
    services: ['chromedriver'],
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    },
}
```

## Android Config

```js
exports.config = {
    specs: [
        './test/specs/**/*.js'
    ],
    exclude: [
        './test/specs/**/example.js'
    ],
    maxInstances: 1,
    port: 4723,
    capabilities: [{
        platformName: 'Android',
    }],
    logLevel: 'info',
    bail: 0,
    baseUrl: 'http://192.168.1.25:8000',
    waitforTimeout: 10000,
    connectionRetryTimeout: 120000,
    connectionRetryCount: 3,
    services: ['appium'],
    framework: 'mocha',
    reporters: ['spec'],
    mochaOpts: {
        ui: 'bdd',
        timeout: 60000
    },
}
```

## Config Diff

```diff
6c6
<         './test/specs/**/*-android.js'
---
>         './test/specs/**/example.js'
8a9
>     port: 4723,
10,11c11
<         browserName: 'chrome',
<         acceptInsecureCerts: true
---
>         platformName: 'Android',
15c15
<     baseUrl: 'http://localhost:8000',
---
>     baseUrl: 'http://192.168.1.25:8000',
19c19
<     services: ['chromedriver'],
---
>     services: ['appium'],
```

## Test Project

[wdio-playground.zip](https://github.com/webdriverio/webdriverio/files/7659906/wdio-playground.zip)

I am attaching my project for reference.  In addition to the aforementioned issue, the project demonstrates that several other basic APIs do _not_ work on Android.  For example:

1. `browser.url('./')` fails on Android.

2. A simple CSS ID selector fails on Android.  `await $('#orderedList')` has to be replaced with `await $('//*[@resource-id="orderedList"]')` on Android.

3. Child xpath works differently. `await list.$$('./*')` on desktop Chrome vs `await list.$$('./*/*')` on Android.

4. Both of the following CSS selectors fail on Android: `$$('ol>li')` and `$$('ol li')`.

5. [getHTML](https://webdriver.io/docs/api/element/getHTML) fails on Android.

Here is how you test with the attached project:

## Running on Desktop Chrome

1. `yarn install`
2. `yarn start` (Equivalent to: `serve -l 8000 html`)
3. `yarn wdio`

## Running on Android

1. `yarn install`
2. `yarn start`. Equivalent to: `serve -l 8000 html`.
3. `start:android`. Equivalent to: `adb shell am start -a android.intent.action.VIEW -d http://192.168.1.25:8000/`.  Of course, please correct the IP address.
4. `yarn wdio wdio.conf.android.js`

**Comment by christian-bromann** [comment]
@salmankhilji can you provide some logs of both test runs, the one run in the browser and the other on mobile?

**Reply by salmankhilji** [reply]
1.  Logs from desktop Chrome—golden path.  Works as expected. 
[wdio-logs-desktop.txt](https://github.com/webdriverio/webdriverio/files/7661299/wdio-logs-desktop.txt)

1.  Logs from Android—broken.
[wdio-logs-android.txt](https://github.com/webdriverio/webdriverio/files/7661302/wdio-logs-android.txt)

**Reply by nic-cross** [reply]
Any update on this issue @christian-bromann ? Thank you.

**Reply by christian-bromann** [reply]
I currently don't have the time to look into this. Any help on this matter would be appreciated.
